### PR TITLE
made checksum tree builder to only use green node not only to cache b…

### PR DESCRIPTION
**Pull request by heejaechang**
…ut to build tree. this should prevent checksum builder to realize all red nodes for solution just to create checksum tree.

this change also include using specialized small dictionary to prevent creating more expensive dictionary.

another optimization can be done is moving checksum data into solution itself so we don't need to have side collections to map between solution green node to checksum node. but I will do that once I get to point where I do more aggressive perf optimizatino to reduce memory foot print if that is necessary

### CODE REVIEW

File: src/Workspaces/Core/Portable/Execution/ChecksumTreeCollection.ChecksumObjectCache.cs [hunk 1]
```diff
@@ -23,40 +20,51 @@ internal partial class ChecksumTreeCollection
         /// </summary>
         private class ChecksumObjectCache
         {
-            private ChecksumObject _checksumObject;
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeCollection.ChecksumObjectCache.cs]: I might just change this to volatile field.

**Comment by heejaechang** [comment]
@mavasani @srivatsn @panopticoncentral @dotnet/roslyn-ide can you take a look?

**Comment by CyrusNajmabadi** [comment]
I can't imagine the memory savings would justify it. At worst don't we only have a checksum per document?  That's going to be a tiny dictionary even in large solutions.

**Comment by heejaechang** [comment]
there is one map for a document. in roslyn there is about 10,000 documents so, it is 10,000 dictionaries.

**Comment by heejaechang** [comment]
ping?

**Comment by heejaechang** [comment]
retest windows_debug_unit64_prtest please

**Comment by heejaechang** [comment]
issue opened - https://github.com/dotnet/roslyn/issues/13243

**Comment by CyrusNajmabadi** [comment]
Why not have one dictionary keyed by documents?

**Comment by CyrusNajmabadi** [comment]
So 10k dictionaries seems rather small.  Say, worst case, a dictionary was somehow 100 bytes**.  That would only be 1 MB of space used.

** How much do you need to store per document? What are you storing per document?

**Comment by heejaechang** [comment]
I updated changeset removing the optimization part. I will address those in other checkin later. this only contains moving builder to green node.

**Comment by heejaechang** [comment]
ping? @CyrusNajmabadi @mavasani @panopticoncentral @srivatsn

File: src/Workspaces/Core/Portable/Execution/ChecksumTreeCollection.ChecksumObjectCache.cs [hunk 2]
```diff
@@ -23,40 +20,41 @@ internal partial class ChecksumTreeCollection
...
+            private ChecksumObject _checksumObject1;
```
Comment (srivatsn) [inline src/Workspaces/Core/Portable/Execution/ChecksumTreeCollection.ChecksumObjectCache.cs]: Rename to _singleElementChecksumObject?

**Comment by srivatsn** [comment]
👍

**Pull request merged**